com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/verifier/JarVerifier.java

```diff
@@ -0,0 +1,162 @@
+        try (InputStream inputStream = jarFile.getInputStream(jarEntry)) {
+            while (inputStream.read(buffer, 0, buffer.length) != -1) {
+                // Just read. This will throw a SecurityException if a signature/digest check fails.
+            }
+        }
+    }
+
+    private void verifyJarEntries(List<X509Certificate> targetCertificates, List<JarEntry> jarEntries, String certificateCN) {
+        for (JarEntry jarEntry : jarEntries) {
+            Certificate[] entryCertificates = jarEntry.getCertificates();
+            if (entryCertificates == null || entryCertificates.length == 0) {
+                verifyIsMetaInformation(jarEntry);
+                continue;
+            }
+            verifyJarEntryIsSigned(targetCertificates, entryCertificates, certificateCN);
+        }
+    }
+
+    private void verifyIsMetaInformation(JarEntry jarEntry) {
+        if (!jarEntry.getName()
+                     .startsWith(META_INF)) {
+            throw new SLException(Messages.ARCHIVE_CONTAINS_UNSIGNED_FILES);
+        }
+    }
+
+    private void verifyJarEntryIsSigned(List<X509Certificate> targetCertificates, Certificate[] entryCertificates, String certificateCN) {
+        List<X509Certificate> entryCertificatesList = convertToX509Certificates(entryCertificates);
+        if (!validCertificateChain(targetCertificates, entryCertificatesList, certificateCN)) {
+            throw new SecurityException(Messages.ARCHIVE_IS_NOT_SIGNED_BY_TRUSTED_SIGNER);
+        }
+    }
+
+    private List<X509Certificate> convertToX509Certificates(Certificate[] entryCertificates) {
+        return Arrays.stream(entryCertificates)
+                     .map(certificate -> (X509Certificate) certificate)
+                     .collect(Collectors.toList());
+    }
+
+    private boolean validCertificateChain(List<X509Certificate> targetCertificates, List<X509Certificate> entryCertificateChain,
+                                          String certificateCN) {
+        boolean certificateCNMatches = false;
+        boolean targetCertificatesMatchEntryCertificate = false;
+        for (X509Certificate entryCertificate : entryCertificateChain) {
+            checkValidityOfCertificate(entryCertificate);
+            if (!certificateCNMatches) {
+                certificateCNMatches = checkCertificateCN(entryCertificate, certificateCN);
+            }
+            if (!targetCertificatesMatchEntryCertificate) {
+                targetCertificatesMatchEntryCertificate = doTargetCertificatesMatchEntryCertificate(targetCertificates, entryCertificate);
+            }
+        }
+        return certificateCNMatches && targetCertificatesMatchEntryCertificate;
+    }
+
+    private void checkValidityOfCertificate(X509Certificate x509Certificate) {
+        try {
+            x509Certificate.checkValidity();
+        } catch (CertificateExpiredException | CertificateNotYetValidException e) {
+            throw new SecurityException(e);
```

Comment:
The Javadoc of SecurityException does not match this usage.

Reply:
Used SLException instead of SecurityException